Shift editor control to TypeScript

abdnh opened the pull request:
Closes https://github.com/ankitects/anki/issues/3830

abdnh committed:
Add ts/routes/editor

abdnh committed:
Move editor code to ts/routes

abdnh committed:
Fix scss paths

abdnh committed:
Merge branch 'main' into editor-3830

abdnh committed:
Move some JS calls in loadNote()

abdnh committed:
Fix encodeIriPaths() not allowed

abdnh committed:
Move screen-specific functionality back NoteEditor

abdnh committed:
Remove console.log() statement

abdnh committed:
Move note saving to TS

abdnh committed:
Fix error on late editor_update_note() call

abdnh committed:
Remove editorReady()

bridgeCommand() looks cleaner here

abdnh committed:
Remove legacy editor tagedit functions

abdnh committed:
Replace note save call for saveTags cmd

abdnh committed:
Fix anki-editable styles not applying

abdnh committed:
Reimplement editor_will_munge_html callbacks

abdnh committed:
Move setSticky() call

abdnh committed:
Move fontMungeHack()

abdnh committed:
Merge branch 'main' into editor-3830

abdnh committed:
Move setColorButtons()

abdnh committed:
Fix color hint not changing

Due to IDs such as `mdi-color-helper` being removed.

abdnh committed:
Move setTagsCollapsed()

abdnh committed:
Move setMathjaxEnabled()

abdnh committed:
Fix wrong exception being catched in get_config_json()

abdnh committed:
Move setShrinkImages()

abdnh committed:
Move setCloseHTMLTags()

abdnh committed:
Formatting

abdnh committed:
Fix lint errors

abdnh committed:
Use orjson helpers

abdnh committed:
Merge branch 'main' into editor-3830

abdnh committed:
Move setupMaskEditor() calls

abdnh committed:
Move showDupes()

abdnh committed:
Move fields_check()

abdnh committed:
Remove legacy routines moved to JS

abdnh committed:
Add a temporary workaround for addcards

abdnh committed:
Fix lint errors

abdnh committed:
Move IO/cloze button toggles

abdnh committed:
Start work on copy & paste handling

abdnh committed:
Use wantsExtendedPaste() for drop event

abdnh committed:
Use wantsExtendedPaste() for paste event

abdnh committed:
Handle internal pastes

abdnh committed:
Move more image occlusion calls

abdnh committed:
Fix lint errors

phu54321 commented:
Some tricky case you might consider?

https://www.gdcdental.com/files/product/Westcott-Curved-%23-S35.jpg

because `Editor._retrieveURL` correctly downloads the URL but unquotes `%23` to `#` upon return, and just assigning `Westcott-Curved-#-S35.jpg` to `img['src']` won't work, as `#` act as a hash tag. It seems like it's the same on your svelte code too?

abdnh committed:
Move context menu logic

abdnh committed:
Move Attach Media button

abdnh committed:
Move Record Audio button

abdnh committed:
Remove mediaPromise

abdnh committed:
Remove legacy editor code

We can worry about add-on compatibility later

abdnh committed:
Fix lint errors

abdnh committed:
Merge branch 'main' into editor-3830

abdnh committed:
Fix Cards/Fields buttons

abdnh committed:
Formatting

abdnh committed:
Remove editor.note refs and fix some lint errors

abdnh committed:
Add missing editor.nid initialization

abdnh committed:
Reimplement sticky field saving

abdnh committed:
Focus field input on drop

abdnh committed:
Fix lint error

abdnh commented:
@dae This is ready for some testing. Pinging @glutanimate @iamllama or anyone interested in giving feedback.

## Known issues

- [ ] Sticky fields not being sticky.
- [ ] Misaligned sticky badge.
- [ ] Fields cleared when switching note type in Add screen.
- [ ] Incomplete pre-add check (`noteCanBeAdded()`)
- [ ] Undo operations for backend calls moved to TypeScript.
- [ ] Toggling MathJax preview doesn't take effect until editor is reopened: https://github.com/ankitects/anki/pull/2014
- [ ] "Paste Image from Clipboard" is less reliable - it doesn't recognize files copied from Explorer on Windows but does work with images copied directly from image viewers or the web.
- [ ] `navigator.clipboard.read()` is a new API, so it needs testing on different platforms.
- [ ] editor.parentWindow.activateWindow() removed from select_image_and_occlude and similar. I assume this is needed for macOS?
- [ ] Svelte implementations for Deck and Notetype selectors.
- [ ] The elephant in the room: Add-on compatibility.

GithubAnon0000 reviewed (commented):
I just build it and tried it. So far, it seems to work quite well (ignoring the issues you already stated). Other issues I found:
1. It was requested that the field text becomes selectable and, ideally, that buttons won't be part of the selection when pressing <kbd>Ctrl+A</kbd>. The first one is an easy fix and I opened https://github.com/ankitects/anki/pull/4122/ to solve it; for the second one, I have no idea. I initially thought your PR was about the review screen (no idea how that happened), so I opened above PR. Maybe it's better if I close it and you change it in your code?
2. The _Eraser_ dropdown menu has `hr`s after every entry. This is a bit too much and could be removed or designed differently.
3. Same as number 2, but for the settings wheel next to _bold_ button.
4. The *Tags* field is cut off a bit at the bottom (maybe an issue with responsive design? Try to reduce the window size in different ways, and you'll see that it's not always cutoff by a lot, and sometimes cut off at the side as well).
5. Related to 4: the scrollbar is cut off as well.
6. The tags don't stay sticky after adding a card.
7. After adding a card, the _History_ button in the bottom right corner doesn't update, meaning it stays grayed out.
8. If you open the _add_ dialog, then immediately click _close_, you'll get a popup asking to discard current input. There is no input though.
9. The right click menu, where _paste_ resides, needs styling to be consistent with ankis ui.
10. I never used this before, so am not sure whether this PR introduced it. If you click a deck to study, then click the *Edit* button during review on a card with an image, then right click said image in the editor and click on _Show in Folder_, the file is opened in the systems webbrowser instead (I would expect it to open the file manager). It does the same like the other option, _Open Image_.
11. Related to 10: If you copy the image via right click, _Copy Image_, the image isn't added to the clipboard. It does work, though, if you <kbd>Ctrl+a</kbd> on the image field and then <kbd>Ctrl+V</kbd> (but not via right-click paste menu).


That's all I could find.

abdnh committed:
Merge branch 'main' into editor-3830

abdnh commented:
Thank you for testing @GithubAnon0000. Regarding some issues:

- (1) I think let's discuss & handle that in your PR, since the changes here are unlikely to touch that part.
- (4) This might be fixed after #4115.
- (5) Hmm, can you see this after pulling the latest changes?
- (10) Let me know if you can reproduce it in the stable version. I suspect it's a system-specific issue. What's your OS?
- (11) What's the format of the image?

user1823 commented:
> But if you can call a command, then xdg-open /path/to/directory will work regardless of desktop environment / regardless of the default file manager.

But, I can't test this. Maybe you can try your suggestion and submit a PR to fix it in the Python code? After looking through the changes in this PR, I realized that if the Python code is made to work on Linux, no additional changes will be required here.

The relevant Python code is here:
https://github.com/ankitects/anki/blob/630bdd31893d5113b81cd60612cef3ebf146d5b4/qt/aqt/editor.py#L1738-L1741
https://github.com/ankitects/anki/blob/630bdd31893d5113b81cd60612cef3ebf146d5b4/qt/aqt/utils.py#L927-L941

iamllama commented:
Thanks for this, it's going to make working on the editor much faster! Noticed a few additional issues as well:
1. The add window now flashes like the stats window, maybe setting the webview bg colour to canvas would help here
2. The "Copy", "Cut" and "Paste" context menu actions don't seem to work on win 10 or wsl for me
3. On dark mode, the context menu's text is white
4. "Discard changes" shows when trying to close after adding a note
5. "Create copy" seems to be broken as well, but differently depending on the notetype
    - Cloze notes: clicking in and out of a field shows an error. On the console it prints `JS error /_app/immutable/chunks/BjIIu02o.mjs:2 Uncaught (in promise) Error` (no AnkiWebView Inspector :/)
    - IO notes: the editor loads and shows the masks/image, but is otherwise not functional (can't change tool)
    - Other types: fields are just blank
6. When adding cloze notes without any clozes, the `You have a cloze notetype but have not made any cloze deletions. Proceed?` dialog doesn't appear anymore. Clicking on the `Add` button just does nothing until a cloze is added
7. "The first field is empty" dialog doesn't show either (related to 6.)
8. When in the browser with a selected row and holding "Down", an error dialog is sometimes shown. When closed (or when the "Down" key is released if no error), the editor appears to rapidly cycle through each note before stopping at the last row when instead it should be debounced


Haven't been able to go through all the commits yet, will get to it when possible!

tatsumoto-ren commented:
>the performance shouldn't be too bad

If you compare webview windows with pure-Qt windows in Anki, you'll notice that Qt windows appear instantly, while webviews can take nearly half a second, or even close to a full second, to load. The worst example is the "deck options" dialog, which used to perform much better in the past and had a more minimalist design.

![screenshot-2025-06-26-20-53-25](https://github.com/user-attachments/assets/2fadfa71-c203-4252-9d2b-ee9a1c0c20da)

I think we can now confidently say that continuing to rely on webviews will negatively impact performance and the overall user experience.

phu54321 commented:
Hope one day we get a pure electron-based webapp of Anki that we can also access on https://ankiweb.net/

abdnh committed:
Merge branch 'main' into editor-3830

abdnh committed:
Add missing allowedSuffixes arg

abdnh committed:
Handle image occlusion drops

abdnh committed:
Trim leading/trailing whitespace from URL

Windows likes to add extra newlines...

abdnh committed:
Go back to using QClipboard

abdnh committed:
Update uv.lock

abdnh committed:
Fix lint errors

abdnh committed:
Merge branch 'main' into editor-3830

abdnh committed:
Formatting

abdnh committed:
Add a hide_while_preserving_layout() call

This should reduce flicker.

abdnh committed:
Use CamelCase consistently for Protobuf names

abdnh committed:
Fix duplicate triggerPageAction() call

abdnh committed:
Fix context menu input field focus handling

phu54321 commented:
If this code moves all editor-related code to typescript, what API's will be exposed to allow custom buttons?

```py
def add_buttons(buttons: List[str], editor: Editor) -> None:
    shortcut = "Ctrl+Alt+Shift+H"

    def _(editor: Editor) -> None:
        editor.web.eval("setFormat('inserthtml', '<img src=_cha_cha-enable.png>');")

    buttons.append(
        editor.addButton(
            icon=None,
            cmd=f"add_cloze_hide_all_marker",
            func=_,
            tip="Make this note like 'Cloze (Hide All)'",
            label="█ CHA",
            keys=shortcut,
        )
    )


gui_hooks.editor_did_init_buttons.append(add_buttons)
```

For example, how should I port this code?

abdnh commented:
@phu54321 editor.addButton() will keep working. Most editor hooks are broken in this PR right now though, and we still need to think about reducing add-on breakages (and probably introduce a JS API).

abdnh committed:
Merge branch 'main' into editor-3830

abdnh committed:
Fix lint errors

abdnh committed:
Add a basic fieldsAreBlank()

dae commented:
Very excited to see the progress here @abdnh! Hope to have this properly reviewed by the end of the week.

> That'll probably never happen given AnkiWeb is strategically designed to not blow up AnkiMobile sales.

While there is an aspect of that, another significant concern is who will pay for the server resources. CPU or I/O-costly features cost us nothing to run on user's own computers, but that's not the case for a hosted service.

abdnh commented on qt/aqt/mediasrv.py:
```diff
@@ -599,6 +602,210 @@ def handle_on_main() -> None:
     return b""
 
 
+def editor_update_note() -> bytes:
```
mediasrv is a bit of a mess now with all this screen-specific logic. Will look into a cleaner solution, maybe by allowing screens to register their own handlers.

dae replied:
AnkiMobile currently does this - each screen registers only the endpoints that are needed, so other endpoints give an error. As we get further along in our porting and we start transitioning from one 'page' to another in the same webview session, it'll probably become less practical to do that, though.

abdnh committed:
Fix misaligned sticky badge

abdnh committed:
Rewrite empty check

abdnh committed:
Fix undo ops of add_note/update_notetype

abdnh committed:
Tweak naming

abdnh committed:
Merge branch 'main' into editor-3830

abdnh committed:
Replace svelte-contextmenu with a custom implementation

Keyboard navigation is to be handled.

abdnh committed:
Implement close button

abdnh committed:
Implement help button

abdnh committed:
Implement add button

dae commented:
Sorry it's taken me a while to circle back to this Abdo. I'm going to start reading through the changes and start testing on iOS so I can get up to speed with things.

> The elephant in the room: Add-on compatibility.

We'll of course want to do what we can here, but I presume there'll be some workflows that can't be easily adapted. I think our best path forward here might be to have a transitionary period where the old and new screens are both available and can be switched between, like as talked about on https://github.com/ankitects/anki/issues/2474#issuecomment-1510775549. That will allow us to iterate on this after the initial merge without any immediate impact on users, and reduces the pressure on us to get this perfect from the start.

The changes in editor.py look fairly extensive, and it may be easiest to copy in a version of the file prior to the changes. For the other files like editcurrent.py, having separate code paths for the two cases might be simplest? For the TS side, you probably have a better idea of whether it would make sense to clone some of the files, the entire ts/editor, or add separate code paths to the existing files. WDYT?

It will be a bit of a pain to make both the old and new approach work at once, and will leave the codebase in a messier state. I'm hoping that it's something we'd have in place for <6 months, not for years like the old deck options/stats screens, as it's too much code to try and maintain multiple copies of for long periods.

Preserving the git history of your changes is more important than the legacy codepath, so I think it might make more sense to put any copies of the old code we make in a new file, rather than copying your changes into a new file (then importing them into the old location so add-ons still find things at the old location). But other ideas welcome.

Will trickle in some more comments as I go through this.

abdnh commented on qt/aqt/mediasrv.py:
```diff
@@ -779,3 +1070,6 @@ def _extract_dynamic_get_request(path: str) -> DynamicRequest | None:
         return None
+        return None
+        return None
+        return None
```
Blame Cursor suggestions 😅 It sometimes decides to duplicate the last line in the file for some reason.

abdnh committed:
Remove unintended change in mediasrv.py

abdnh committed:
Remove old requirements.txt files

tatsumoto-ren commented:
Since these changes negatively impact the user experience on desktop, it would be better if they were implemented in an add-on instead. Alternatively, there should be a way to completely and permanently opt out of them. I'm not a fan of transition periods, as they don’t solve the problem, they only postpone it.

abdnh commented:
> That will allow us to iterate on this after the initial merge without any immediate impact on users, and reduces the pressure on us to get this perfect from the start.

That's reasonable. I was worried figuring out add-on compatibility would take a lot of time and discussions.


> The changes in editor.py look fairly extensive, and it may be easiest to copy in a version of the file prior to the changes. For the other files like editcurrent.py, having separate code paths for the two cases might be simplest? For the TS side, you probably have a better idea of whether it would make sense to clone some of the files, the entire ts/editor, or add separate code paths to the existing files. WDYT?

I agree it will be simpler to have separate Python files for the new code. For the TS side, most of my changes consist of replacing bridgeCommand() calls with a backend call and some logic ported from Python. As there's not much change to the code flow, I feel the simplest option to add a legacy switch to existing code to make it go back to bridgeCommand handling.

phu54321 commented:
This looks like the old Anki window.

<img width="970" height="1079" alt="image" src="https://github.com/user-attachments/assets/dba8813b-3744-4668-9aa2-0a415481f22b" />

I guess this PR is to resolve future technical debt, but I think this could be re-implementing some other WYSIWYG toolkits like [summernote](https://summernote.org/) or [tinymce](https://www.tiny.cloud/). 

I once used Summernote for making [kian](https://github.com/phu54321/kian), so maybe you can look at it. Better to avoid reinventing the wheels?

abdnh committed:
Merge branch 'main' into editor-3830

abdnh committed:
Add back legacy code as separate screens

abdnh committed:
Remove duplicate lines in mediasrv.py again

Turned out to be a race condition when the ruff/black/isort extensions
are installed.

https://github.com/astral-sh/ruff-vscode/issues/651

abdnh committed:
Restore old editor entrypoint

abdnh committed:
Add a legacy switch to NoteEditor

abdnh committed:
Fix empty field's height in legacy editor

abdnh committed:
Add missing await

abdnh committed:
Implement History button

abdnh committed:
Set a different shortcut for Mac

abdnh committed:
Remove Qt buttons from new Add screen

abdnh committed:
Fix Create Copy

abdnh committed:
Remove unused add_current_note()

abdnh committed:
Fix bottom editor buttons not shown in IO editor

Positioning needs to be fixed.

abdnh committed:
Add hack for buttons position in IO mode

abdnh committed:
Merge branch 'main' into editor-3830

abdnh committed:
Fix Toast component mounting

abdnh committed:
Hide IO toast on note change

abdnh committed:
Debounce loadNote() calls

abdnh committed:
Fix lint errors

abdnh committed:
Complete noteCanBeAdded()

abdnh committed:
Merge branch 'main' into editor-3830

abdnh committed:
Reduce duplication across modal components

abdnh committed:
Fix Modal event handlers not working

abdnh committed:
Fix TextInputModal not closing when OK is clicked

abdnh committed:
Implement Deck/Notetype selectors in Svelte

abdnh committed:
Disable prefer-const for Svelte for now

https://github.com/sveltejs/eslint-plugin-svelte/issues/818#issuecomment-2570986849

abdnh committed:
Use translations

abdnh committed:
Pass selected notetype/deck to editor

abdnh committed:
Only show selectors in Add mode

abdnh committed:
Add selector shortcuts

abdnh committed:
Add selector tooltips

abdnh committed:
Do not show selectors in legacy mode

abdnh committed:
Fix lint errors

abdnh committed:
Use defaults_for_adding()

abdnh committed:
Add missing loadNewNote arg

abdnh committed:
Use default_deck_for_notetype()

abdnh committed:
Remove legacy addcards selectors from new screen

abdnh committed:
Fix selected notetype/deck reverting back to defaults

abdnh committed:
Only reload note on reopen if empty

abdnh committed:
Do not fire initial onChange

This causes the note type to revert back to the first item after initial
note load

abdnh committed:
Fix default_deck_for_notetype handling

abdnh committed:
Move defaultsForAdding handling to start of loadNote

abdnh committed:
Clear lastAddedNote on notetype change

abdnh committed:
Merge branch 'main' into editor-3830

abdnh committed:
Implement field copying on notetype change

abdnh committed:
Formatting

abdnh committed:
Simplify loadNote() arg passing

abdnh committed:
Fix wrong arg name

abdnh committed:
Alow passing deck/notetype ID on AddCards init

abdnh committed:
Fix TS errors

abdnh committed:
Ensure bigint

abdnh committed:
Fix lints

abdnh committed:
Copy sticky fields

abdnh committed:
Focus selector's search input

abdnh committed:
Clear search on selector toggle

abdnh committed:
Preserve focused field

abdnh committed:
Fix browser error on deleted notes

abdnh committed:
Remove repeated line

abdnh committed:
Fix getCard() not allowed

abdnh committed:
Fix legacy DuplicateLink handling

abdnh committed:
Merge branch 'main' into editor-3830

abdnh committed:
Show number of added cards

abdnh committed:
Merge branch 'main' into editor-3830

abdnh committed:
Move retrieve_url to backend

abdnh committed:
Merge branch 'main' into editor-3830

abdnh committed:
Add setupColourPalette() to legacy editor

abdnh committed:
Restore toggleMathjax for legacy editor

abdnh committed:
Add back .empty MathJax styles

abdnh committed:
Refresh fields after MathJax toggle

There's probably a more efficient solution for this but I couldn't
figure it out

abdnh committed:
Handle undo for all RPCs that return OpChanges

abdnh committed:
Remove unused CSS selector

abdnh committed:
Merge branch 'main' into editor-3830

abdnh committed:
Handle operation changes from other screens

abdnh committed:
Fix notetype label not changing when default deck is set

abdnh committed:
Fix undo of methods returning OpChanges directly not being handled correctly

abdnh committed:
Move .reload_note() calls

abdnh committed:
Formatting

abdnh committed:
Remove log()

abdnh committed:
Handle deeply nested OpChanges

Messages such as AddNotesResponse were not handled correctly.

abdnh committed:
Run scheduled save before add

abdnh committed:
Save before opening Fields/Cards

abdnh committed:
Add RPCs for Fields/Cards dialogs

abdnh committed:
Fix call_after_note_saved()

abdnh committed:
Fix legacy editor not rebuilding after changes in ts/routes

abdnh committed:
Fix legacy editable styles

abdnh committed:
Merge branch 'main' into editor-3830

abdnh committed:
Fix formatting/typing errors

abdnh committed:
Simplify async request handling

abdnh committed:
Allow opening legacy screens from browser/reviewer

abdnh committed:
Fix deck of copied note not being used

abdnh committed:
Fix TS type

abdnh committed:
Make CooldownTimer accept an aync action

abdnh committed:
Preserve initial note load args

abdnh committed:
Fix notetype/deck of original note not being set

abdnh committed:
Avoid extra initial loadNote()

abdnh committed:
Fix focused field not being preserved

abdnh committed:
Formatting

abdnh committed:
Fix double paste

abdnh committed:
Fix disabled history button's hover styles

The `background` property from the primary styles had a higher priority

abdnh committed:
Fix initial=True being passed to subsequent loadNote calls

abdnh committed:
Fix notetype changing after changes to fields

abdnh committed:
Scroll selected deck/notetype into view

abdnh committed:
Keep selector's search bar in view

abdnh committed:
Formatting

abdnh committed:
Fix notetype ID not being set after note added

abdnh committed:
Merge branch 'main' into editor-3830

abdnh committed:
Work around bottom toolbar being cut off in IO editor

We should look into a better CSS layout in the future

abdnh committed:
Remove unused assert

abdnh committed:
Abort note loading if note switched quickly

See https://forums.ankiweb.net/t/preload-images-in-the-editor/67795/5?u=abdo

The issue might still be triggered in rare cases. We should probably
combine all backend calls into a single abortable call.

abdnh committed:
Merge branch 'main' into editor-3830

abdnh committed:
Fix toast message only shown on first added note

dae commented:
This is an impressive amount of work Abdo, and I'm sorry I let this stall for so long.

Currently this uses the new editor by default, which is probably going to be disruptive. We might want to consider defaulting to the old editor instead (and making sure both addcards,editcurrent and browser all support the shift toggle).

Couple of minor issues I noticed while testing:
- editor_did_init is called twice (by __init__ and a later editorReady)
- "1 card added" got stuck for me when I added the second or third card, and did not go away.
<img width="706" height="131" alt="Screenshot 2026-04-25 at 20 10 51" src="https://github.com/user-attachments/assets/00228916-08e6-4df2-b987-f9fe8fe4c074" />
- Minor tweaks to layout/formatting to make things look neater - I think Luc may have brought that up above too.

tatsumoto-ren commented:
How about releasing this as an add‑on first, and if it gains enough popularity, consider adding it to the main program? I think it could cause too many breakages if it's merged directly. Also, you wouldn't gain anything from including it in Anki right now, since the functionality already exists in the current implementation.

abdnh committed:
Fix Toast not hiding after first add

abdnh committed:
Remove extra editor_did_init call

abdnh committed:
Default to legacy editor

abdnh committed:
Merge branch 'main' into editor-3830

abdnh committed:
Fix formatting

abdnh committed:
Add back legacy editor handling to browser

abdnh committed:
Fix legacy IO addition path broken

abdnh committed:
Remove unused ts/editor/NoteEditor.svelte

Accidentally added back by a merge commit

abdnh committed:
Improve layout of IO editor

Fix suggested by Claude. Needs testing on mobile.

andrewsanchez committed:
Merge branch 'main' into editor-3830

fernandolins committed:
build(ts): add root tsconfig for Playwright @generated path alias

fernandolins committed:
test(e2e): expand editor fixtures with bridgeStub, editorPage and editor

fernandolins committed:
test(e2e): add shared Playwright helpers for editor e2e tests

fernandolins committed:
test(e2e): add note-add roundtrip suite

Verifies the new SvelteKit editor sends a well-formed addNote RPC:
- Types into two fields with pressSequentially() (fires real key
  events for shadow-DOM contenteditables).
- Decodes /_anki/addNote payload and asserts field values + deckId.
- Confirms updateNotes never fires (add-mode contract).
- Asserts /_anki/newNote fires after Add (form reset signal).
- Checks field 0 clears and bridgeCommand("saved") was emitted.

fernandolins committed:
test(e2e): add sticky field toggle suite

Verifies the non-legacy sticky path (StickyBadge.svelte) uses
getNotetype + updateNotetype RPCs instead of bridgeCommand:
- Targets .field-state [role="button"] to skip the CollapseLabel
  span (which also carries role="button").
- Decodes updateNotetype payload, asserts fields[0].config.sticky.
- Checks "highlighted" class and absence of toggleSticky bridge call.
- Toggles back to restore the notetype to its original state.

fernandolins committed:
test(e2e): add HTML paste-filter suite

End-to-end coverage for ts/lib/html-filter running on real paste:
- Case 1: <p> tags in pasted HTML are rewritten to <div> by
  convertToDiv (element.ts:59); asserts innerHTML and addNote payload.
- Case 2: <script> tags are stripped and never execute; asserts
  innerHTML, window.__xssRan sentinel stays false, and payload is clean.
Uses pasteData() to construct the ClipboardEvent inside the browser
so DataTransfer is a real browser API instance.

fernandolins committed:
test(e2e): harden RPC matchers and protobuf error messages in helpers

fernandolins committed:
Fixes lint issue on fixtures

fernandolins committed:
test(e2e): harden note-add roundtrip assertions

fernandolins committed:
test(e2e): enforce legacy parity for pasted <script> content

- Use isRpc() predicate consistently in both waitForRequest and
  waitForResponse matchers
- Assert script content does not leak as visible text in the field,
  enforcing parity with the legacy BeautifulSoup .decompose() path ‚Äî
  test intentionally fails until element.ts maps SCRIPT to removeElement

fernandolins committed:
test(e2e): use isRpc() predicate in sticky-field RPC matchers

fernandolins committed:
test(e2e): add edge cases and parity gaps for editor suite

- Assert empty first field blocks addNote RPC (QT parity: noteCanBeAdded)
- Assert sticky field value survives add, non-sticky field clears
  (QT parity: _get_sticky_fields)
- Assert onclick/onerror attributes stripped from pasted HTML and never
  execute (QT parity: filterAttributes allowlist in element.ts)

fernandolins committed:
Fixes lint issue

fernandolins committed:
test(e2e): add editor parity coverage for focused add and script paste

fernandolins committed:
test(e2e): add editor chooser and RPC helpers

fernandolins committed:
test(e2e): cover editor add context switching parity

abdnh committed:
Merge remote-tracking branch 'origin' into editor-3830

abdnh committed:
Update test to reflect p > div conversion

`<p>` to `<div>` conversion used to be handled on the Python side.

abdnh committed:
Fix legacy editor using new copy/paste handling

abdnh committed:
Completely strip script tags instead of unwrapping

Matches the behavior of _pastePreFilter

abdnh committed:
Update script stripping test

abdnh committed:
Set CSP directives for SvelteKit pages

Prevents user script execution in the editor.

fernandolins committed:
test(e2e): expand context-switching coverage to address reviewer feedback

- Rename "notetype persists" test to "notetype and deck persist" and
  switch the add to use TEST_DECK instead of Default so the assertion
  exercises the real _nt_{ntid}_lastDeck persistence path rather than
  the trivially-satisfied Default fallback.

- Add setConfigJson helper and force Mode B (addToCur=false) in
  beforeEach: fresh Anki collections default to Mode A (addToCur=true
  per schema11), which caused all Mode B assertions to silently pass
  or fail against the wrong code path.

- Add test: mode B session — switching notetype auto-selects the last
  deck used with that notetype (onNotetypeChange → defaultDeckForNotetype
  → deckChooser.select, no user interaction required).

- Add test: mode A reopen — last notetype used for the current deck is
  restored via _deck_{did}_lastNotetype.

- Add test: mode B fallback (no history) — when _nt_{ntid}_lastDeck is
  absent or points to a deleted deck, defaults_for_adding falls back to
  the collection's curDeck, not Default.

- Add test: mode A fallback — when _deck_{did}_lastNotetype is absent,
  default_notetype_for_deck falls back to get_current_notetype_for_adding.

All branches of rslib/src/adding.rs are now covered.

fernandolins committed:
chore(mediasrv): expose config and deck management RPCs

Add set_config_bool, set_config_json_no_undo, remove_config and
set_current_deck to the exposed_backend_list so they are reachable
via /_anki/{method} from the web layer.

These RPCs are already defined in the proto and have _raw variants in
the backend; exposing them is consistent with the existing set_config_json
and get_config_bool entries.

fernandolins committed:
Fixes lint issue

fernandolins committed:
test(e2e): group context-switching tests into logical describe blocks

abdnh committed:
Revert accidental submodule change

abdnh committed:
Fix error on opening Stats screen

Luc-Mcgrady commented:
The large margin is caused by the margin and spacing of the "verticalLayout_3" widget in the UI_Dialog 
https://github.com/ankitects/anki/blob/71ae5a6b673bc8943f048fed906751febfe585bd/qt/aqt/forms/addcards.ui#L26
QT:<img width="827" height="69" alt="image" src="https://github.com/user-attachments/assets/af691e7f-952b-4291-a055-94821b2ec4d5" />Typescript:<img width="827" height="91" alt="image" src="https://github.com/user-attachments/assets/051c6805-e1ea-4272-9ce9-2d0e7519ce9e" />With the spacing removed it typescript looks like QT:
<img width="837" height="69" alt="image" src="https://github.com/user-attachments/assets/b1c43e64-619a-4173-8d73-2b91c19be4e6" />

You'll have to fix it programmatically somehow  because this change also affects the QT reviewer :( 

Also if I open "edit current" (pressing e inside the reviewer) without holding shift it is broken. I get the error "unexpected editor type: review"

Hope this helps~

abdnh committed:
Merge pull request #10 from fernandolins/test/e2e-editor-suite

test(e2e): add Playwright e2e suite for Svelte editor

abdnh committed:
Fix error on opening legacy EditCurrent

abdnh committed:
Reduce top spacing in NewAddCards

abdnh commented:
@tatsumoto-ren 

> I think it could cause too many breakages if it's merged directly.

It's disabled by default and will be put behind a preferences toggle (#4871) soon while we look into add-on compatibility and the long-term plan.

> Also, you wouldn't gain anything from including it in Anki right now, since the functionality already exists in the current implementation.

We can at least get some beta testing to catch regressions and maybe test it on AnkiMobile/AnkiDroid.

fernandolins committed:
Removes unnecessarily exposed endpoints

abdnh committed:
Merge pull request #11 from fernandolins/test/e2e-context-switching

test(e2e): add e2e coverage for editor add context switching
